src/search.cpp

```diff
@@ -211,18 +200,70 @@ void Search::clear() {
   Tablebases::init(Options["SyzygyPath"]); // Free mapped files
 }
 
+/// MainThread::perft() divide the perft search evenly through Threads
+void MainThread::perft() {
+
+  StateInfo st;
+  std::string posNow = rootPos.fen();
+  std::vector<perftPos> perftRoot, newPerftRoot;
+
+  // Step 1: Create new root positions that will be distributed between threads
+  for (const auto& m : MoveList<LEGAL>(rootPos)) {
+      rootPos.do_move(m, st);
+      perftSum[perftRoot.size()].store(0, std::memory_order_release);
+      perftRoot.push_back({rootPos.fen(), Depth(Limits.perft - 1), (int)perftRoot.size()});
+      rootPos.undo_move(m);
+  }
+
+  // Step 2: Extend root positions if it's not enough to distribute evenly between threads
+  while (perftRoot.size() && perftRoot[0].d > 3 && 8 * Threads.size() > perftRoot.size()){
+      newPerftRoot.clear();
+      for (const auto& pm : perftRoot){
+          rootPos.set(pm.fen, rootPos.is_chess960(), rootPos.state(), this);
+
+          for (const auto& m : MoveList<LEGAL>(rootPos)){
+              rootPos.do_move(m, st);
+              newPerftRoot.push_back({rootPos.fen(), pm.d - 1, pm.ID});
+              rootPos.undo_move(m);
+          }
+      }
+      perftRoot = newPerftRoot;
+  }
+
+  // Step 3: Share root positions evenly between the threads
+  unsigned cntT = Threads.size(); auto pr = perftRoot.begin();
+  for (Thread* th : Threads){
+      const unsigned Total = (perftRoot.end() - pr) / (cntT--);
+      th->rootPerft = new perftPos[Total + 1];
+
+      for (unsigned i = 0; i < Total; i++)
+          th->rootPerft[i] = *pr++;
+      th->rootPerft[Total] = {"", 0, -1};
+  }
+
+  // Step 4: Initialize Perft in all threads and wait all to be finished
+  Threads.start_searching();
+  Thread::search();
+  Threads.wait_for_search_finished();
+
+  // Step 5: Gather information between threads and print results
+  int i = 0; nodes = 0;
```

Comment:
I don't like how these statements are on the same line. It's slightly confusing.

Reply:
Sure, I can make them in different lines.